models/object_detection_nanodet/demo.cpp

```diff
@@ -0,0 +1,520 @@
+        exp(input.reshape(0, input.total() / (this->regMax + 1)), x_exp);
+        reduce(x_exp, x_sum, 1, REDUCE_SUM, CV_32F);
+        repeat(x_sum, 1, this->regMax + 1, x_repeat_sum);
+        divide(x_exp, x_repeat_sum, result);
+        return result;
+    }
+
+    Mat applyProjection(Mat& input) 
+    {
+        Mat repeat_project;
+        repeat(this->project, input.rows, 1, repeat_project);
+        multiply(input, repeat_project, input);
+        reduce(input, input, 1, REDUCE_SUM, CV_32F);
+        Mat projection = input.col(0).clone();
+        return projection.reshape(0, projection.total() / 4);
+    }
+
+    void preNMS(Mat& anchors, Mat& bbox_pred, Mat& cls_score, const int nms_pre = 1000)
+    {
+        Mat max_scores;
+        reduce(cls_score, max_scores, 1, REDUCE_MAX);
+
+        Mat indices;
+        sortIdx(max_scores.t(), indices, SORT_DESCENDING);
+
+        Mat indices_float = indices.colRange(0, nms_pre);
+        Mat selected_anchors, selected_bbox_pred, selected_cls_score;
+        for (int j = 0; j < indices_float.cols; ++j) 
+        {
+            selected_anchors.push_back(anchors.row(indices_float.at<int>(j)));
+            selected_bbox_pred.push_back(bbox_pred.row(indices_float.at<int>(j)));
+            selected_cls_score.push_back(cls_score.row(indices_float.at<int>(j)));
+        }
+
+        anchors = selected_anchors;
+        bbox_pred = selected_bbox_pred;
+        cls_score = selected_cls_score;
+    }
+
+    void clipBoundingBoxes(Mat& x1, Mat& y1, Mat& x2, Mat& y2) 
+    {
+        Mat zeros = Mat::zeros(x1.size(), x1.type());
+        x1 = min(max(x1, zeros), Scalar(this->imageShape.width - 1));
+        y1 = min(max(y1, zeros), Scalar(this->imageShape.height - 1));
+        x2 = min(max(x2, zeros), Scalar(this->imageShape.width - 1));
+        y2 = min(max(y2, zeros), Scalar(this->imageShape.height - 1));
+    }
+
+    Mat calculateBoundingBoxes(const Mat& anchors, const Mat& bbox_pred) 
+    {
+        Mat x1 = anchors.col(0) - bbox_pred.col(0);
+        Mat y1 = anchors.col(1) - bbox_pred.col(1);
+        Mat x2 = anchors.col(0) + bbox_pred.col(2);
+        Mat y2 = anchors.col(1) + bbox_pred.col(3);
+
+        clipBoundingBoxes(x1, y1, x2, y2);
+
+        Mat bboxes;
+        hconcat(vector<Mat>{x1, y1, x2, y2}, bboxes);
+
+        return bboxes;
+    }
+
+    vector<Rect2d> bboxMatToRect2d(const Mat& bboxes)
+    {
+        Mat bboxes_wh(bboxes.clone());
+        bboxes_wh.colRange(2, 4) = bboxes_wh.colRange(2, 4) -= bboxes_wh.colRange(0, 2);
+        vector<Rect2d> boxesXYXY;
+        for (size_t i = 0; i < bboxes_wh.rows; i++) 
+        {
+            boxesXYXY.emplace_back(bboxes.at<float>(i, 0),
+                                   bboxes.at<float>(i, 1),
+                                   bboxes.at<float>(i, 2),
+                                   bboxes.at<float>(i, 3));
+        }
+        return boxesXYXY;
+    }
+
+    std::tuple<Mat, Mat> getClassIdAndConfidences(const Mat& scores)
```

Comment:
I propose to use C-style function signature so that we don't need tuple:

```cpp
void getClassAndConfidences(const Mat &scores, Mat &classIds, Mat &confidences)
```

Or simply extract this function.

Reply:
Extracted! I attempted to refactor it but have ended up making it less concise.